consensus/src/pending_order_votes.rs

```diff
@@ -0,0 +1,262 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_consensus_types::{common::Author, order_vote::OrderVote};
+use aptos_crypto::{hash::CryptoHash, HashValue};
+use aptos_logger::prelude::*;
+use aptos_types::{
+    aggregate_signature::PartialSignatures,
+    ledger_info::{LedgerInfo, LedgerInfoWithPartialSignatures, LedgerInfoWithSignatures},
+    validator_verifier::{ValidatorVerifier, VerifyError},
+};
+use std::{collections::HashMap, sync::Arc};
+
+/// Result of the order vote processing. The failure case (Verification error) is returned
+/// as the Error part of the result.
+#[derive(Debug, PartialEq, Eq)]
+pub enum OrderVoteReceptionResult {
+    /// The vote has been added but QC has not been formed yet. Return the amount of voting power
+    /// QC currently has.
+    VoteAdded(u128),
+    /// The very same vote message has been processed in past.
+    DuplicateVote,
+    /// The very same author has already voted for another proposal in this round (equivocation).
+    EquivocateVote,
+    /// This block has just been certified after adding the vote.
+    NewLedgerInfoWithSignatures(Arc<LedgerInfoWithSignatures>),
+    /// There might be some issues adding a vote
+    ErrorAddingVote(VerifyError),
+    /// Error happens when aggregating signature
+    ErrorAggregatingSignature(VerifyError),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum OrderVoteStatus {
+    EnoughVotes,
+    NotEnoughVotes,
+}
+
+/// A PendingVotes structure keep track of order votes for the last few rounds
+pub struct PendingOrderVotes {
+    /// Maps LedgerInfo digest to associated signatures (contained in a partial LedgerInfoWithSignatures).
+    /// Order vote status stores caches the information on whether the votes are enough to form a QC.
+    li_digest_to_votes: HashMap<
+        HashValue, /* LedgerInfo digest */
+        (OrderVoteStatus, LedgerInfoWithPartialSignatures),
+    >,
+    /// Map of (Author, round number) to order vote. This is useful to discard multiple votes.
+    author_to_order_vote: HashMap<(Author, u64), OrderVote>,
```

Comment:
I don't think we need this, we can attach a QC to the corresponding order vote and verify them so it can't equivocate here. also it helps avoid a DDoS attack that can OOM this struct by sending future rounds

Reply:
Are you suggesting to add QC as a new field in OrderVote? If the receiver has to verify QC for each of the order votes, that's an additional cost. And what if a validator receives order vote before getting the block, the validator won't be able to verify QC.

Reply:
not as new field to OrderVote, but OrderVoteMsg, yes there's additional verification but probably negligible. Verifying a QC doesn't require having the block, it's just signature verification. also we can't vote for it without the block anyway